### PR TITLE
Adding ability to tell that a section variable has to be hidden by default at goal initialization

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** developer-side feature 

This is in preparation of #12432, in relation to this [comment](https://github.com/coq/coq/pull/12432#issuecomment-664005730) and comments before and after.

The PR is at the current stage to get feedback on the "proof of concept" (starting with @ejgallego, specialist of `declare.ml` and @ppedrot, who expressed a related need).

### CODE REVIEW

File: vernac/locality.ml [hunk 1]
```diff
@@ -11,7 +11,8 @@
 (** * Managing locality *)
 
 type import_status = ImportDefaultBehavior | ImportNeedQualified
-type locality = Discharge | Global of import_status
+type goal_visibility = bool
+type locality = Discharge of goal_visibility | Global of import_status
```
Comment (SkySkimmer) [inline vernac/locality.ml:15]:
```suggestion
type locality = Discharge of { goal_visibility : bool } | Global of import_status
```

**Comment by SkySkimmer** [comment]
>in relation to this comment

If we want letins that get inlined by discharge the info would need to be in the kernel, not in decls AFAICT

File: vernac/declare.ml [hunk 2]
```diff
@@ -1404,8 +1404,10 @@ let initialize_named_context_for_proof () =
   List.fold_right
     (fun d signv ->
       let id = NamedDecl.get_id d in
-      let d = if Decls.variable_opacity id then NamedDecl.drop_body d else d in
-      Environ.push_named_context_val d signv) sign Environ.empty_named_context_val
+      if Decls.variable_goal_visibility id then
+        let d = if Decls.variable_opacity id then NamedDecl.drop_body d else d in
+        Environ.push_named_context_val d signv
+      else signv) sign Environ.empty_named_context_val
```
Comment (SkySkimmer) [inline vernac/declare.ml:1410]: If we don't put ring's lemmas in the context ring can't use them, so I'm not sure how it's supposed to work.

**Comment by SkySkimmer** [comment]
Before looking at the PR contents I was expecting a modification of Context.Named to add a visibility field to each declaration, with the field only impacting printing (and maybe inlining at discharge). Did you consider doing that?
There would be some irritating effects eg doing `intro ring_lemma_1` would fail even though the name appears to be free so that direction is not without issues.

File: interp/decls.ml [hunk 3]
```diff
@@ -62,6 +62,7 @@ type logical_kind =
 type variable_data = {
   opaque:bool;
   kind:logical_kind;
+  goal_visibility:bool;
```
Comment (ejgallego) [inline interp/decls.ml:65]: I am not sure this table is the right place to store this info, instead it seems to me that all of `Decls` should go away.

If the data has a lifetime similar to `Proof.t` info should live there I guess.

**Comment by herbelin** [comment]
> Before looking at the PR contents I was expecting a modification of `Context.Named` to add a visibility field to each declaration, with the field only impacting printing (and maybe inlining at discharge). Did you consider doing that?

Yes. I did not work at the level of the kernel because it seems to me that the consensus was not yet found about how to think at `named_context`. One direction which I conceptually like is that section variables eventually have a qualified name so that the name of a section variable can be reused in the goal context w/o preventing all global definitions depending on the section variable with same name to be used, or without `Show Proof` getting too confused, etc. Then, this global name would be converted to a `Rel` at discharge time. This is not w/o adaptation though, since `mkVar` would sometimes have to mean `mkConst` of the section variable (if ever `mkConst` is used to represent section variables). Or, alternatively, the argument of `mkVar` could be changed to take either a section variable or a goal variable and `named_context` would be a list of such declarations with names being either a section or a goal variable. Maybe the latter would be the easiest way to make a transition.

> There would be some irritating effects eg doing `intro ring_lemma_1` would fail even though the name appears to be free so that direction is not without issues.

Which would be fixed if we explicitly distinguish between goal and section variables though.

**Comment by herbelin** [comment]
PS: Considering #12890, I'm leaning towards eventually closing this PR, targetting instead a deeper evolution of `named_context`.

**Comment by SkySkimmer** [comment]
>PS: Considering #12890, I'm leaning towards eventually closing this PR, targetting instead a deeper evolution of named_context.

I agree

**Pull request closed without merging**